lib/rmg/TreasurePlacer.cpp

```diff
@@ -0,0 +1,795 @@
+
+void TreasurePlacer::process()
+{
+	addAllPossibleObjects();
+	auto * m = zone.getModificator<ObjectManager>();
+	if(m)
+		createTreasures(*m);
+}
+
+void TreasurePlacer::init()
+{
+	DEPENDENCY(ObjectManager);
+	DEPENDENCY(ConnectionsPlacer);
+	POSTFUNCTION(RoadPlacer);
+}
+
+void TreasurePlacer::setQuestArtZone(Zone * otherZone)
+{
+	questArtZone = otherZone;
+}
+
+void TreasurePlacer::addAllPossibleObjects()
+{
+	ObjectInfo oi;
+	
+	int numZones = static_cast<int>(map.getZones().size());
+	
+	for(auto primaryID : VLC->objtypeh->knownObjects())
+	{
+		for(auto secondaryID : VLC->objtypeh->knownSubObjects(primaryID))
+		{
+			auto handler = VLC->objtypeh->getHandlerFor(primaryID, secondaryID);
+			if(!handler->isStaticObject() && handler->getRMGInfo().value)
+			{
+				for(auto temp : handler->getTemplates())
+				{
+					if(temp.canBePlacedAt(zone.getTerrainType()))
+					{
+						oi.generateObject = [temp]() -> CGObjectInstance *
+						{
+							return VLC->objtypeh->getHandlerFor(temp.id, temp.subid)->create(temp);
+						};
+						auto rmgInfo = handler->getRMGInfo();
+						oi.value = rmgInfo.value;
+						oi.probability = rmgInfo.rarity;
+						oi.templ = temp;
+						oi.maxPerZone = rmgInfo.zoneLimit;
+						vstd::amin(oi.maxPerZone, rmgInfo.mapLimit / numZones); //simple, but should distribute objects evenly on large maps
+						possibleObjects.push_back(oi);
+					}
+				}
+			}
+		}
+	}
+	
+	if(zone.getType() == ETemplateZoneType::WATER)
+		return;
+	
+	//prisons
+	//levels 1, 5, 10, 20, 30
+	static int prisonsLevels = std::min(generator.getConfig().prisonExperience.size(), generator.getConfig().prisonValues.size());
+	for(int i = 0; i < prisonsLevels; i++)
+	{
+		oi.generateObject = [i, this]() -> CGObjectInstance *
+		{
+			std::vector<ui32> possibleHeroes;
+			for(int j = 0; j < map.map().allowedHeroes.size(); j++)
+			{
+				if(map.map().allowedHeroes[j])
+					possibleHeroes.push_back(j);
+			}
+			
+			auto hid = *RandomGeneratorUtil::nextItem(possibleHeroes, generator.rand);
+			auto factory = VLC->objtypeh->getHandlerFor(Obj::PRISON, 0);
+			auto obj = (CGHeroInstance *) factory->create(ObjectTemplate());
+			
+			
+			obj->subID = hid; //will be initialized later
+			obj->exp = generator.getConfig().prisonExperience[i];
+			obj->setOwner(PlayerColor::NEUTRAL);
+			map.map().allowedHeroes[hid] = false; //ban this hero
+			generator.decreasePrisonsRemaining();
+			obj->appearance = VLC->objtypeh->getHandlerFor(Obj::PRISON, 0)->getTemplates(zone.getTerrainType()).front(); //can't init template with hero subID
+			
+			return obj;
+		};
+		oi.setTemplate(Obj::PRISON, 0, zone.getTerrainType());
+		oi.value = generator.getConfig().prisonValues[i];
+		oi.probability = 30;
+		oi.maxPerZone = generator.getPrisonsRemaning() / 5; //probably not perfect, but we can't generate more prisons than hereos.
+		possibleObjects.push_back(oi);
+	}
+	
+	//all following objects are unlimited
+	oi.maxPerZone = std::numeric_limits<ui32>().max();
```

Comment:
In other places we've got default values of -1 **and** 1, which is risky at least.

Reply:
Do you propose to use -1 instead?